paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+            Lower limit of the range where alpha factors are drawn from
+
+        All other arguments are described in "calc_rirs".
+
+    Returns:
+        (rir, alpha):
+        rir: None, if no solution could be found for given parameters
+        otherwise combined and scaled RIR
+        alpha: the directivity factor of acoustic source used for generation
+    """
+    # maximum peak of pra is shifted by 40 samples compared to physical
+    # time of arrival (TOA)
+    pra_toa_offset = 40
+
+    # calculate stochastic reverberation part of RIR
+    n = np.arange(rir_len)
+    delta = 3 * np.log(10) / t60
+    toa_direct = int(
+        np.round(distance / sound_speed * sample_rate)) + pra_toa_offset
+    raw_stochastic = np.random.normal(size=rir_len) * np.exp(
+        -delta * (n - toa_direct) / sample_rate)
+    raw_stochastic[:toa_direct + 1] = 0
+
+    lower_lim = toa_direct - win_len_direct
+    upper_lim = toa_direct + win_len_direct + 1
+
+    # adjust alpha to maintain solvability;
+    # in some cases, the RIR of image source method may be already too diffuse
+    # according to target DRR for the currently drawn alpha, so adding
+    # additional reverberant parts causes an invalid solution
+    # if change_alpha is set after some iterations of rejection sampling,
+    # calculate alpha based on the early reflections of the RIR that will
+    # guarantee a solution during scaling to match the target DRR,
+    # but still matches the original range of alpha the best
+    if change_alpha:
+        # calculate DRR of ISM part
+        drr_im = (np.sum(rir_ism[lower_lim:upper_lim] ** 2)
+                  / np.sum(rir_ism[upper_lim:] ** 2))
+        # solve DRR formula to get alpha that would just create the
+        # DRR of the early reflections of the RIR, seen as upper bound for the
+        # maximum DRR to achieve, because only first reflections are simulated
+        # and amending of the stochastic part would overall lower the DRR
+        max_factor = (drr_im * np.pi * t60 * distance ** 2
+                      / np.prod(room_dimensions) * 100
+                      / directivity_attenuation ** 2)
+        if max_factor < alpha:
+            # for current alpha, target DRR would be more direct as DRR of
+            # early reflections, so re-draw alpha, that then creates a lower,
+            # but achievable target DRR and therefore a solution during
+            # later scaling can be found
+            if max_factor < lower_limit_alpha:
+                # use omnidirectional directivity factor (1) as lower limit
+                # because otherwise no solution can be found
+                alpha = np.random.uniform(1, max_factor)
+            else:
+                # sample alpha until valid solution found, that still lays
+                # within the original range, but preserves convergence
+                alpha = np.random.uniform(lower_limit_alpha, max_factor)
+                while alpha > max_factor:
```

Comment:
Why do you need the loop? If you draw with  the value of alpha with np.random.uniform(lower_limit_alpha, max_factor)  the resulting value of alpha definitely is smaller than max_factor.